### PR TITLE
MONIT-32321: Updating AWS files to ingest logs in proxy

### PR DESCRIPTION
**Pull request by deosu**
https://jira.eng.vmware.com/browse/MONIT-32321
https://confluence.eng.vmware.com/pages/viewpage.action?pageId=1574459690

### CODE REVIEW

File: aws/src/index.js [hunk 1]
```diff
@@ -349,9 +350,37 @@ class S3HttpCollector extends Collector {
...
+  } else if (logStream.startsWith('i-')) {
+    const logStreamKeys = logGroup.split('-');
+    source = logStreamKeys[logStreamKeys.length - 1];
+  }
+
   for (const logEvent of cloudWatchLogs.logEvents) {
     logEvent.log_type = 'aws_cloud_watch';
+    logEvent.source = source;
+    logEvent.accoundId = accountId;
+    logEvent.Region = region;
```
Comment (prabhupo) [inline aws/src/index.js:378]: logEvent.Region -> logEvent.region, can we use 'r'?

File: aws/src/index.js [hunk 2]
```diff
@@ -349,9 +350,37 @@ class S3HttpCollector extends Collector {
...
+  const logStream = cloudWatchLogs.logStream;
+
+  if(logGroup.startsWith('/aws/lambda/')) {
+    const logGroupKeys = logGroup.split('/');
+    source = logGroupKeys[logGroupKeys.length - 1];
+  } else if (logStream.startsWith('i-')) {
+    const logStreamKeys = logGroup.split('-');
+    source = logStreamKeys[logStreamKeys.length - 1];
+  }
+
```
Comment (prabhupo) [inline aws/src/index.js:373]: Can this be part of utility method to derive source?

File: aws/src/lint.js [hunk 3]
```diff
@@ -51,18 +52,59 @@ class LIntHttpEnv {
...
-    const body = [];
-    res.on('data', chunk => body.push(chunk));
-    res.on('end', () => {
-      if (res.statusCode < 200 || res.statusCode >= 300) {
-        reject(new Error(`statusCode: ${res.statusCode}, response: ${Buffer.concat(body).toString()}`));
-      } else {
-        resolve('Log sent to Log Intelligence!');
-      }
-    });
+class LIntWFProxyEnv {
```
Comment (prabhupo) [inline aws/src/lint.js:55]: Is creating explicit LIntWFProxyEnv related class necessary? Lambda should be agnostic of presence of Proxy. All it needs to know is whether to call http vs https end point, pass query params or not etc.
Can't we make changes without this class creation?

File: aws/src/lint.js [hunk 4]
```diff
@@ -51,18 +52,59 @@ class LIntHttpEnv {
...
+  return https.request(options, (res) => {
+    processResult(res, reject, resolve);
+  });
+}
+
+const executeRequest = (collector, postData) => new Promise((resolve, reject) => {
+  const options = collector.lintEnv.createRequestOptions(collector.structure);
+  let req;
+
+  if (collector.lintEnv instanceof LIntWFProxyEnv) {
```
Comment (prabhupo) [inline aws/src/lint.js:103]: Can't we detect the protocol by looking at the endpoint?

File: aws/templates/aws-template-cli.yaml [hunk 5]
```diff
@@ -8,7 +8,7 @@ Parameters:
         Default: ""
     APIUrl:
         Type: "String"
-        Description: "Data Url of vRealize Log Insight Cloud"
+        Description: "Data Url of vRealize Log Insight Cloud or Wavefront proxy(Note: APIToken is not required for Wavefront proxy URL)"
```
Comment (prabhupo) [inline aws/templates/aws-template-cli.yaml:11]: Need Amit K to review this description.

File: aws/templates/aws-template.yaml [hunk 6]
```diff
@@ -8,7 +8,7 @@ Parameters:
         Default: ""
     APIUrl:
         Type: "String"
-        Description: "Data Url of vRealize Log Insight Cloud"
+        Description: "Data Url of vRealize Log Insight Cloud or Wavefront proxy(Note: APIToken is not required for Wavefront proxy URL)"
```
Comment (prabhupo) [inline aws/templates/aws-template.yaml:11]: Need Amit K to review this description.

Comment (prabhupo) [reply aws/templates/aws-template.yaml:11]: Also this commit doesn't have changes around API Token validation. So isn't this description inaccurate?

**Review by prabhupo — COMMENTED** [review]
1. Did we run 'yarn test' on the changes?
2. Can we add unit tests for the incoming changes?

File: aws/src/index.js (hunk 1 above)
Comment (deosu) [reply aws/src/index.js]: To correlate logs with metrics we will need to add a tag `Region`(with `R`). Cloud metrics add this tag as well.

File: aws/src/index.js (hunk 2 above)
Comment (deosu) [reply aws/src/index.js]: Agreed. Made the necessary changes.

File: aws/src/lint.js (hunk 3 above)
Comment (deosu) [reply aws/src/lint.js]: Agreed. Refactor the code.

File: aws/templates/aws-template-cli.yaml (hunk 5 above)
Comment (deosu) [reply aws/templates/aws-template-cli.yaml]: I will share the draft PR with Amit K.

File: aws/templates/aws-template.yaml (hunk 6 above)
Comment (deosu) [reply aws/templates/aws-template.yaml]: I will share the draft PR with Amit K.

**Comment by deosu** [comment]
Added code coverage for index.js changes.

File: aws/src/index.js [hunk 7]
```diff
@@ -349,33 +348,54 @@ class S3HttpCollector extends Collector {
...
   }
+return source;
+}
+/* eslint-disable no-param-reassign */
+const processLogText = (cloudWatchLogs, tagRegexMap, arn) => {
+  // derive accountId & Region
+  const { region, accountId } = deriveRegionAndAccount(arn);
+
+  // derive source: default
+  let source = accountId + '-' + region;
```
Comment (prabhupo) [inline aws/src/index.js:381]: Minor: lets move this into deriveSource() method.

File: aws/src/index.js [hunk 8]
```diff
@@ -349,33 +348,54 @@ class S3HttpCollector extends Collector {
...
-  // derive source for lambda function or EC2 instance
+  if (arn) {
+    const keys = arn.split(':');
+    region = keys[3];
+    accountId = keys[4];
+  }
+return {region, accountId};
+}
+function deriveSource(cloudWatchLogs,source){
   const logGroup = cloudWatchLogs.logGroup;
```
Comment (prabhupo) [inline aws/src/index.js:363]: Just want to confirm: logGroup, logStream is set for the batch not in individual log level right?

File: aws/src/lint.js [hunk 9]
```diff
@@ -84,28 +65,23 @@ function processResult(res, reject, resolve) {
...
+const sendHttpRequest = (options, postData) => new Promise((resolve, reject) => {
+  const req = http.request(options, (res) => {
     processResult(res, reject, resolve);
   });
-}
-
-const executeRequest = (collector, postData) => new Promise((resolve, reject) => {
-  const options = collector.lintEnv.createRequestOptions(collector.structure);
-  let req;
+  req.on('error', error => reject(error));
```
Comment (prabhupo) [inline aws/src/lint.js:72]: Line#72 - lines#76 are duplicate of lines#85 - lines#93. Can we move these lines into a helper method? Say have processResult() as common method to process response from sendHttpsRequest(), sendHttpRequest() methods. Let sendHttpsRequest/sendHttpRequest make https/http calls and in turn call processResult().

File: aws/src/index.js (hunk 8 above)
Comment (deosu) [reply aws/src/index.js:363]: The fields `logGroup`, & `logStream` are part of the input CloudWatch Log. That is why they will show up on a log message sent to the proxy. But we do not add these tags to individual `logEvents`.

File: aws/src/index.js (hunk 7 above)
Comment (deosu) [reply aws/src/index.js:381]: Sure. Will do.

File: aws/test/__snapshots__/cloudwatch_collector.test.js.snap [hunk 10]
```diff
@@ -5,8 +5,11 @@ Object {
   "ingest_timestamp": 1538769915817,
   "logEvents": Array [
     Object {
+      "Region": "region",
+      "accountId": "accountId",
       "id": "id1",
       "log_type": "aws_cloud_watch",
+      "source": "accountId-region",
```
Comment (prabhupo) [inline aws/test/__snapshots__/cloudwatch_collector.test.js.snap:12]: Minor: 'accountId-region' -> should this be 'accountId-Region' as we are using 'R' in region?

File: aws/test/__snapshots__/cloudwatch_collector.test.js.snap [hunk 11]
```diff
@@ -23,8 +26,11 @@ Object {
...
```
Comment (prabhupo) [inline aws/test/__snapshots__/cloudwatch_collector.test.js.snap:33]: Minor: 'accountId-region' -> should this be 'accountId-Region' as we are using 'R' in region?

File: aws/src/lint.js (hunk 9 above)
Comment (deosu) [reply aws/src/lint.js:72]: Sure. Will make the change.

File: aws/src/index.js [hunk 12]
```diff
@@ -349,33 +348,54 @@ class S3HttpCollector extends Collector {
...
   const logStream = cloudWatchLogs.logStream;
 
-  if(logGroup.startsWith('/aws/lambda/')) {
+  if(logGroup && logGroup.startsWith('/aws/lambda/')) {
     const logGroupKeys = logGroup.split('/');
     source = logGroupKeys[logGroupKeys.length - 1];
-  } else if (logStream.startsWith('i-')) {
+  } else if (logStream && logStream.startsWith('i-')) {
     const logStreamKeys = logGroup.split('-');
```
Comment (prabhupo) [inline aws/src/index.js]: logGroup.split('-'); -> should this be logStreamKeys.split('-');

**Pull request closed without merging**